common/src/app/common/pages/changes.cljc

```diff
@@ -494,3 +452,36 @@
   (ex/raise :type :not-implemented
             :code :operation-not-implemented
             :context {:type (:type op)}))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Component changes detection
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Analyze one change and checks if if modifies the main instance of
+;; any component, so that it needs to be synced immediately to the
+;; main component. Return the ids of the components that need sync.
+
+(defmulti components-changed (fn [_ change] (:type change)))
+
+(defmethod components-changed :mod-obj
+  [file-data {:keys [id page-id _component-id operations]}]
+  (when page-id
+    (let [page (ctpl/get-page file-data page-id)
+          shape-and-parents (map #(ctn/get-shape page %)
+                                 (into [id] (cph/get-parent-ids (:objects page) id)))
+          need-sync? (fn [operation]
+                       ; We need to trigger a sync if the shape has changed any
+                       ; attribute that participates in components syncronization.
+                       (and (= (:type operation) :set)
+                            (component-sync-attrs (:attr operation))))
+          any-sync? (some need-sync? operations)]
+      (when any-sync?
+        (into #{} (->> shape-and-parents
+                       (filter #(:main-instance? %)) ; Select shapes that are main component instances
+                       (map :id)))))))
```

Comment:
Use transducers here.
Additionally: there are no need for anonymous function on `(filter #(:main-instance? %))` -> it can be directly `(filter :main-instance?)`

Reply:
done